### ISSUE TITLE
docker: downstream containers rebuild too frequently

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#483]
# Problem description
The garage codebase is copied in `Dockerfile.base`, which means that any docker-compose stack with `Dockerfile.base` upstream will rebuild all downstream containers every time the code is changed.

For example, if I use the `-headless` image, it has a few extra packages on top of `-base`.

If I execute the sequence
```sh
make run-headless RUN_CMD="python examples/tf/ppo_pendulum.py"  # builds garage-base followed by garage-headless
# edit examples/tf/ppo_pendulum.py in some trivial way
make run-headless RUN_CMD="python examples/tf/ppo_pendulum.py"  # rebuilds last layer of garage-base followed by all of garage-headless
```

This is a minor annoyance for small layers like `-headless` but could be a huge problem for larger downstream layers (e.g. a ROS container).

# Suggested remedies
1. Only copy files as-necessary in garage-base and move the final codebase copy to downstream containers
1. Use an FS mount for the code (this is messier but makes it easier to inherit from downstream containers -- e.g. if I tried to layer a ROS container on top of garage-headless I'd have the same problem all over again)
1. Split major downstream containers into 2 layers: a middle layer which installs deps (e.g. `-headless` and `-nvidia` and then a common runtime container which copies the code (e.g. a new stack would be `garage-base`-->`garage-headless`-->`garage-runtime` and `garage-base`-->`garage-nvidia`-`garage-runtime`)
#

# Code for context
from Dockerfile.base
```Dockerfile
# ...

# Add code stub last
COPY . /root/code/garage
RUN ["/bin/bash", "-c", "source activate garage && pip install -e /root/code/garage && source deactivate garage"]
```

from Dockerfile.headless
```Dockerfile
ARG PARENT_IMAGE=rlworkgroup/garage-base
FROM $PARENT_IMAGE

# apt dependencies
RUN \
  apt-get -y -q update && \
  # Prevents debconf from prompting for user input
  # See https://github.com/phusion/baseimage-docker/issues/58
  DEBIAN_FRONTEND=noninteractive apt-get install -y \
    # Dummy X server
    xvfb \
    pulseaudio && \
  apt-get clean && \
  rm -rf /var/lib/apt/lists/*

# Ready, set, go.
ENTRYPOINT ["docker/entrypoint-headless.sh"]
```

**Comment by gonzaiva** [comment]
Yes, this is a major problem indeed, specially when dealing with the garage-ros-intera containers, since the ros layers will be rebuilt if the local repository of the garage is modified.
I noticed you've already addressed suggestion 1 in #487, but as you mentioned, this solution won't work for the garage-ros-intera containers, so between 2 and 3, I prefer 2 to avoid adding more image layers. The only detail is that we're already adding a volume for the experiment data with the container name in it, so if we link the entire local repository now, we have to find another way to redirect the data to the specific container folder.

**Comment by ryanjulian** [comment]
I think 3 is the probably the most Docker-friendly option and it will be invisible to almost all users (except for those seeking to implement custom containers).

Making an FS mount for code isn't very Docker-ly.

**Comment by gonzaiva** [comment]
Haha Docker-ly, okay, I will work on number 3 to solve this issue

**Comment by ryanjulian** [comment]
Don't clear your schedule for it. I already cleared the immediate hassle for most garage users with my PR. I guess you can just slip it in to support the ROS docker containers.

**Comment by ryanjulian** [comment]
@gonzaiva can you verify the new design fixes this?

**Comment by gonzaiva** [comment]
I tested the three containers, making modifications to the local files in each of them, and only the last layer (runtime-layer) is rebuilt, so no other task is ran again to re-build the image. It's fixed  with #557.
